### PR TITLE
Updated type defs for @testing-library/user-event

### PR DESCRIPTION
**Pull request by weyert**
Please fill in this template.

- [ ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [X] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [X] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [X] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/testing-library/user-event/releases/tag/v4.1.0
- [X] Increase the version number in the header if appropriate.
- [X] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by johnnyreilly** [comment]
Heya!

Looks like there's some errors:

```
Error in testing-library__user-event
Error: /home/travis/build/DefinitelyTyped/DefinitelyTyped/types/testing-library__user-event/index.d.ts:6:18
ERROR: 6:18   interface-name  interface name must not have an "I" prefix
ERROR: 18:26  eofline         file should end with a newline
/home/travis/build/DefinitelyTyped/DefinitelyTyped/types/testing-library__user-event/testing-library__user-event-tests.ts:5:1
ERROR: 5:1    expect          TypeScript@next expected type to be:
  void
got:
  any
ERROR: 5:11   expect          TypeScript@next compile error: 
Property 'selectOptions' does not exist on type '{ click: (element: Element | Window) => void; dblClick: (element: Element | Window) => void; type: (element: Element | Window, text: string, userOpts?: UserOpts | undefined) => Promise<...>; }'.
ERROR: 6:1    expect          TypeScript@next expected type to be:
  void
got:
  any
ERROR: 6:11   expect          TypeScript@next compile error: 
Property 'selectOptions' does not exist on type '{ click: (element: Element | Window) => void; dblClick: (element: Element | Window) => void; type: (element: Element | Window, text: string, userOpts?: UserOpts | undefined) => Promise<...>; }'.
```

https://travis-ci.org/DefinitelyTyped/DefinitelyTyped/builds/550307788?utm_source=github_status&utm_medium=notification

File: types/testing-library__user-event/index.d.ts [hunk 1]
```diff
@@ -1,20 +1,18 @@
-// Type definitions for @testing-library/user-event 4.0
+// Type definitions for @testing-library/user-event 4.1
 // Project: https://github.com/testing-library/user-event
 // Definitions by: Wu Haotian <https://github.com/whtsky>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
-export interface UserOpts {
+export interface IUserOptions {
```
Comment (whtsky) [inline types/testing-library__user-event/index.d.ts]: any particular reason for changing the interface's name?

IMO it's a breaking change and the "I" prefix is making tslint unhappy

Comment (weyert) [reply types/testing-library__user-event/index.d.ts]: Yes, sorry, on my computer it was complaining about the I prefix within my project. I have resolved it now :)

Regarding the name just felt better to me but reverted the name change

**Comment by johnnyreilly** [comment]
Okay it looks like I was wrong about the relative import.... Want to substitute for the namespaced one instead?

**Comment by johnnyreilly** [comment]
Maybe it doesn't need the prefix? Not sure...

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/babel-types/index.d.ts

**Comment by johnnyreilly** [comment]
Huh failed. Weird. Maybe we import fron: `testing-library__user-event`?

**Comment by johnnyreilly** [comment]
Or do a `require` like in babel-template: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/babel-template/babel-template-tests.ts

I think that's an example of a namespaced / scoped definition

**Comment by johnnyreilly** [comment]
@andrewbranch @drosenwasser - sorry to bother you but we're bumping up against something I haven't had to do before. Namely, write tests for a scoped / namespaced package. 

We don't seem to be able to import the type definition in our tests. We've tried:

- `import ... from 'user-event'` - this is what is committed now and seems incorrect; it refers to the legacy (pre-scoped) location of the package
- `import ... from './` - this is not allowed
- `import ... from '@testing-library/user-event' - this seems correct but does not work.

I feel like we're missing a trick / technique here and I was wondering if you knew what it was.  I've tried inspecting other scoped packages on DT but the answer has not presented itself...

I suspect we could hack this by just updating the legacy definition of `user-event` but that doesn't seem like the right thing to do 🤗

**Comment by sandersn** [comment]
@johnnyreilly I think you have to put explicit entries in tsconfig. Take a look at `@babel/core` for an example.

**Comment by andrewbranch** [comment]
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/efa6b9e11eb0167547d6611490be48b94bc8d7ea/types/babel__core/tsconfig.json#L18-L31

**Comment by johnnyreilly** [comment]
Aha! The ol' `paths` magic! Thanks @sandersn and @andrewbranch!

@weyert shall we give this a go?

**Review by whtsky — APPROVED** [review]

**Comment by johnnyreilly** [comment]
@sandersn @andrewbranch - unfortunately the paths magic didn't work 😢 

We tried approach suggested:

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/36444/commits/3c9cb6d021e69fb56c91c0cdd1d82d788ae662ea

But we still ended up seeing the `Cannot find module '@testing-library/user-event'` error message:

https://travis-ci.org/DefinitelyTyped/DefinitelyTyped/builds/550712352#L704

@weyert has (for now) switched to use the exisiting approach where we point at the existing `user-event` module.  I'm going to merge so I can unblock him.  

But I'd like to understand why we couldn't get this to work.  As far as I can tell, we did exactly the same approach as the `@babel/core` and yet it didn't work.  I've picked through this line by line but kind see any reason for failure. Do you have any insights?

**Pull request merged**